pkg/dynamiccontroller/dynamic_controller.go

```diff
@@ -306,265 +304,372 @@ func (dc *DynamicController) processNextWorkItem(ctx context.Context) bool {
+	err := handler(ctx, ctrl.Request{NamespacedName: oi.NamespacedName})
 	if err != nil {
 		handlerErrorsTotal.WithLabelValues(gvrKey).Inc()
 	}
 	return err
 }
 
-// shutdown performs a graceful shutdown of the controller.
-func (dc *DynamicController) shutdown(ctx context.Context) error {
-	dc.log.Info("Starting graceful shutdown")
-
-	var wg sync.WaitGroup
-	dc.informers.Range(func(key, value interface{}) bool {
-		k := key.(schema.GroupVersionResource)
-		dc.log.V(1).Info("Shutting down informer", "gvr", k.String())
-		wg.Add(1)
-		go func(informer *informerWrapper) {
-			defer wg.Done()
-			informer.informer.Shutdown()
-		}(value.(*informerWrapper))
-		return true
-	})
-
-	// Wait for all informers to shut down or timeout
-	done := make(chan struct{})
-	go func() {
-		wg.Wait()
-		close(done)
-	}()
-
-	select {
-	case <-done:
-		dc.log.Info("All informers shut down successfully")
-	case <-ctx.Done():
-		dc.log.Error(ctx.Err(), "Timeout waiting for informers to shut down")
-		return ctx.Err()
+func (dc *DynamicController) enqueueParent(parentGVR schema.GroupVersionResource, obj interface{}, eventType string) {
+	mobj, err := meta.Accessor(obj)
+	if err != nil {
+		dc.log.Error(err, "Failed to get meta for object to enqueue", "eventType", eventType)
+		return
 	}
 
-	return nil
-}
+	oi := ObjectIdentifiers{NamespacedName: types.NamespacedName{
+		Namespace: mobj.GetNamespace(),
+		Name:      mobj.GetName(),
+	}, GVR: parentGVR}
+	dc.log.V(1).Info("Enqueueing object", "objectIdentifiers", oi, "eventType", eventType)
 
-// ObjectIdentifiers is a struct that holds the namespaced key and the GVR of the object.
-//
-// Since we are handling all the resources using the same handlerFunc, we need to know
-// what GVR we're dealing with - so that we can use the appropriate workflow operator.
-type ObjectIdentifiers struct {
-	// NamespacedKey is the namespaced key of the object. Typically in the format
-	// `namespace/name`.
-	NamespacedKey string
-	GVR           schema.GroupVersionResource
+	informerEventsTotal.WithLabelValues(parentGVR.String(), eventType).Inc()
+	dc.queue.Add(oi)
 }
 
-// updateFunc is the update event handler for the GVR informers
-func (dc *DynamicController) updateFunc(old, new interface{}) {
-	newObj, ok := new.(*unstructured.Unstructured)
-	if !ok {
-		dc.log.Error(nil, "failed to cast new object to unstructured")
+func (dc *DynamicController) updateFunc(parentGVR schema.GroupVersionResource, oldObj, newObj interface{}) {
+	newMeta, err := meta.Accessor(newObj)
+	if err != nil {
+		dc.log.Error(err, "failed to access new object meta")
 		return
 	}
-	oldObj, ok := old.(*unstructured.Unstructured)
-	if !ok {
-		dc.log.Error(nil, "failed to cast old object to unstructured")
+	oldMeta, err := meta.Accessor(oldObj)
+	if err != nil {
+		dc.log.Error(err, "failed to access old object meta")
 		return
 	}
-
-	if newObj.GetGeneration() == oldObj.GetGeneration() {
+	if newMeta.GetGeneration() == oldMeta.GetGeneration() {
```

Comment:
Is this always correct? Are there cases where annotation or labels change in a way that will result in us wanting to reconcile parent, but that doesn't trigger generation update?

Reply:
I believe that we have 2 cases I know of where we can  have a problem with this, mainly config maps and secrets. pretty much all other resources that have a spec subresource will be able to use this filter on generation.